gcc/rust/feature/rust-feature.h

```diff
@@ -0,0 +1,76 @@
+// Copyright (C) 2020-2022 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+namespace Rust {
+
+struct Feature
+{
+  enum State
+  {
+    ACCEPTED,
+    ACTIVE,
+    REMOVED,
+    STABILIZED,
+  };
+  union Edition
+  {
+    CompileOptions::Edition edition;
+    uint8_t none;
+    Edition (CompileOptions::Edition e) : edition (e) {}
+    Edition (uint8_t n) : none (n) {}
+  };
+  Feature (State S, const char *N, const char *RS, const char *GS, uint64_t I,
+	   Edition E, const char *D)
+    : state (S), name (N), rustc_since (RS), gccrs_since (GS), issue (I),
+      edition (E), description (D)
+  {
+    Feature::NUM++;
```

Comment:
Global state is not great for this. I'd like to rethink that API as I think it does too much. In your current class, a `Feature` represents a singular feature, but you can also call `Feature::load_hash_map()` to get access to all features defined within the compiler. I think this is an indication the class should be split in two:

1. A `Feature` class which should represent a singular feature. So what you have at the moment is good
2. A `FeatureCtx` or `FeatureTable` class which keeps track of which features are present and their state, and maybe keeps track of which features are being *used* in the present crate. But we can think about that later on